### PR TITLE
Fixes tesla coil hell contraptions

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request
Removes the ability for tesla coils to generate power with more then 85% efficiency, fixing #52851. (Yes I tested it this time)
Cleans up the remainder of my zap_act refactor, making the proc better fit its usecase and removing some unneeded code. 
Adds a check in the tesla coil zap() proc that makes sure we're not trying to use power that's not there.
Removes some seemingly complex math from said proc, replaces it with a static 20% draw * the efficiency.

## Why It's Good For The Game

Please stop making infinite energy devices engineers, well at least this one was fun

## Changelog
:cl:
balance: Tesla coils max efficiency rate has been dropped from 400% to 85% to help prevent the below 
fix: The tesla coil can no longer make infinite energy
/:cl:

### CODE REVIEW

File: code/game/machinery/_machinery.dm [hunk 1]
```diff
@@ -577,14 +577,15 @@ Class Procs:
...
-/obj/machinery/zap_act(power, zap_flags, shocked_objects)
-	. = ..()
+/obj/machinery/zap_act(power, zap_flags)
 	if(prob(85) && (zap_flags & ZAP_MACHINE_EXPLOSIVE) && !(resistance_flags & INDESTRUCTIBLE))
 		explosion(src, 1, 2, 4, flame_range = 2, adminlog = FALSE, smoke = FALSE)
 	else if(zap_flags & ZAP_OBJ_DAMAGE)
 		take_damage(power/2000, BURN, "energy")
 		if(prob(40))
 			emp_act(EMP_LIGHT)
+		power -= power / 2000
```
Comment (Rohesie) [inline code/game/machinery/_machinery.dm:587]:
```suggestion
		power -= power * 0.0005
```

File: code/game/machinery/_machinery.dm [hunk 2]
```diff
@@ -577,14 +577,15 @@ Class Procs:
...
+	. = ..(power, zap_flags)
```
Comment (Rohesie) [inline code/game/machinery/_machinery.dm:588]:
```suggestion
	return ..()
```
Arguments are passed by default, and yes, the modifications carry on.

File: code/modules/antagonists/blob/structures/_blob.dm [hunk 3]
```diff
@@ -207,13 +207,13 @@
...
 
-/obj/structure/blob/zap_act(power)
-	. = ..()
+/obj/structure/blob/zap_act(power, zap_flags)
 	if(overmind)
 		if(overmind.blobstrain.tesla_reaction(src, power))
 			take_damage(power/400, BURN, "energy")
 	else
 		take_damage(power/400, BURN, "energy")
+	. = ..(power - (power / 400), zap_flags) //You don't get to do it for free
```
Comment (Rohesie) [inline code/modules/antagonists/blob/structures/_blob.dm:216]:
```suggestion
		take_damage(power/400, BURN, "energy")
	power -= power * 0.0025 //You don't get to do it for free
	return ..()
```

File: code/modules/power/lighting.dm [hunk 4]
```diff
@@ -745,8 +745,7 @@
 	if(zap_flags & ZAP_MACHINE_EXPLOSIVE)
 		explosion(src,0,0,0,flame_range = 5, adminlog = FALSE)
 		qdel(src)
-	else
-		return ..()
+	. = ..(power, zap_flags & ~ZAP_MACHINE_EXPLOSIVE & ~ZAP_OBJ_DAMAGE)
```
Comment (Rohesie) [inline code/modules/power/lighting.dm:748]: Calling parent after deleting is a bad idea.
Also same as the other ones, don't modify the arguments in the parent call, do it in the proc properly.

File: code/modules/power/tesla/coil.dm [hunk 5]
```diff
@@ -31,7 +31,7 @@
 	for(var/obj/item/stock_parts/capacitor/C in component_parts)
 		power_multiplier += C.rating
 		zap_cooldown -= (C.rating * 20)
-	input_power_multiplier = power_multiplier
+	input_power_multiplier = (0.85 * (power_multiplier / 4)) //Max out at 85% efficency.
```
Comment (Rohesie) [inline code/modules/power/tesla/coil.dm:34]:
```suggestion
	input_power_multiplier = power_multiplier * 0.2125 //Max out at 85% efficency.
```

File: code/modules/power/tesla/coil.dm [hunk 6]
```diff
@@ -71,35 +71,34 @@
...
 		var/datum/bank_account/D = SSeconomy.get_dep_account(ACCOUNT_ENG)
 		if(D)
 			D.adjust_money(min(power_produced, 1))
-		addtimer(CALLBACK(src, .proc/reset_shocked), 10)
+		flick("coilhit", src)
 		zap_buckle_check(power)
 		playsound(src.loc, 'sound/magic/lightningshock.ogg', 100, TRUE, extrarange = 5)
-		return power_produced
+		obj_flags |= BEING_SHOCKED
+		addtimer(CALLBACK(src, .proc/reset_shocked), 10)
```
Comment (Rohesie) [inline code/modules/power/tesla/coil.dm:87]:
```suggestion
		addtimer(CALLBACK(src, .proc/reset_shocked), 1 SECONDS)
```

File: code/modules/power/tesla/generator.dm [hunk 7]
```diff
@@ -8,3 +8,4 @@
 /obj/machinery/the_singularitygen/tesla/zap_act(power, zap_flags)
 	if(zap_flags & ZAP_MACHINE_EXPLOSIVE)
 		energy += power
+	. = ..(power, zap_flags & ~ZAP_MACHINE_EXPLOSIVE & ~ZAP_OBJ_DAMAGE) // Don't blow yourself up yeah?
```
Comment (Rohesie) [inline code/modules/power/tesla/generator.dm:11]:
```suggestion
		energy += power
	zap_flags &= ~(ZAP_MACHINE_EXPLOSIVE | ZAP_OBJ_DAMAGE) // Don't blow yourself up yeah?
	return ..()
```

File: code/game/machinery/_machinery.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/game/machinery/_machinery.dm:588]: That's really good to know, thank you

File: code/modules/power/lighting.dm (hunk 4 above)
Comment (LemonInTheDark) [reply code/modules/power/lighting.dm:748]: ahhh gottcha

File: code/modules/power/tesla/coil.dm (hunk 5 above)
Comment (LemonInTheDark) [reply code/modules/power/tesla/coil.dm:34]: It's an operation faster but a lot less clear. I don't think I want to do this

File: code/modules/power/tesla/generator.dm [hunk 8]
```diff
@@ -8,3 +8,5 @@
 /obj/machinery/the_singularitygen/tesla/zap_act(power, zap_flags)
 	if(zap_flags & ZAP_MACHINE_EXPLOSIVE)
 		energy += power
+	zap_flags &= ~(ZAP_MACHINE_EXPLOSIVE | ZAP_OBJ_DAMAGE) // Don't blow yourself up yeah?
+	. = ..()
```
Comment (Rohesie) [inline code/modules/power/tesla/generator.dm:12]:
```suggestion
	return ..()
```
Return at the end of the proc

File: tgstation.dme [hunk 9]
```diff
@@ -1718,6 +1718,7 @@
 #include "code\modules\cargo\exports\sheets.dm"
 #include "code\modules\cargo\exports\tools.dm"
 #include "code\modules\cargo\exports\weapons.dm"
+#include "code\modules\cargo\exports\xenobio.dm"
```
Comment (Rohesie) [inline tgstation.dme:1721]: Intended change?

**Comment by Rohesie** [comment]
Test this locally. If everything is working it's good to go after the change request.

File: tgstation.dme (hunk 9 above)
Comment (LemonInTheDark) [reply tgstation.dme:1721]: Braindamage

File: code/modules/power/tesla/generator.dm (hunk 8 above)
Comment (LemonInTheDark) [reply code/modules/power/tesla/generator.dm:12]: ahhh thanks for catching that

**Comment by LemonInTheDark** [comment]
I've done some testing already, but I'll redo it with the changes to be sure we're all good

**Comment by LemonInTheDark** [comment]
Can confirm, things work. *crosses fingers*

**Pull request merged**